sermant-plugins/sermant-mq-grayscale/sermant-mq-grayscale-plugin/src/main/java/io/sermant/mq/grayscale/config/MqGrayscaleConfig.java

```diff
@@ -0,0 +1,207 @@
+            }
+        }
+        return map;
+    }
+
+    /**
+     * return the traffic tag item based on traffic properties matching result
+     *
+     * @param properties traffic
+     * @return gray tag item
+     */
+    public Optional<GrayTagItem> matchGrayTagByTrafficProperties(Map<String, String> properties) {
+        for (GrayTagItem item : getGrayscale()) {
+            if (!StringUtils.isEmpty(item.trafficMatchProperties(properties))) {
+                return Optional.of(item);
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * return the traffic tag item based on serviceMeta properties matching result
+     *
+     * @param properties serviceMeta
+     * @return gray tag item
+     */
+    public Optional<GrayTagItem> matchGrayTagByServiceMeta(Map<String, String> properties) {
+        for (GrayTagItem item : getGrayscale()) {
+            if (item.serviceMetaMatchProperties(properties)) {
+                return Optional.of(item);
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * return the traffic tag item by grayGroupTag
+     *
+     * @param grayGroupTag grayGroupTag
+     * @return gray tag item
+     */
+    public Optional<GrayTagItem> getGrayTagByGroupTag(String grayGroupTag) {
+        for (GrayTagItem item: getGrayscale()) {
+            if (grayGroupTag.equals(item.getConsumerGroupTag())) {
+                return Optional.of(item);
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * build traffic tag properties to string
+     *
+     * @return traffic tag string
+     */
+    public String buildAllTrafficTagInfoToStr() {
+        StringBuilder sb = new StringBuilder();
+        for (GrayTagItem item : getGrayscale()) {
+            if (sb.length() > 0) {
+                sb.append(SubscriptionDataUtils.AFA_SYMBOL);
+            }
+            sb.append(item.getConsumerGroupTag());
+            for (Map.Entry<String, String> entry : item.getTrafficTag().entrySet()) {
+                sb.append(entry.getKey())
+                        .append(SubscriptionDataUtils.AFA_SYMBOL)
+                        .append(entry.getValue());
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * compare source/target MqGrayscaleConfig excludeGroupTags config whether to change
+     *
+     * @param target MqGrayscaleConfig
+     * @return isChanged
+     */
+    public boolean isBaseExcludeGroupTagsChanged(MqGrayscaleConfig target) {
+        HashSet<String> targetBaseExcludeTags = new HashSet<>(target.getBase().getExcludeGroupTags());
+        HashSet<String> sourceBaseExcludeTags = new HashSet<>(getBase().getExcludeGroupTags());
+        targetBaseExcludeTags.removeAll(sourceBaseExcludeTags);
+        target.getBase().getExcludeGroupTags().forEach(sourceBaseExcludeTags::remove);
+        return !targetBaseExcludeTags.isEmpty() || !sourceBaseExcludeTags.isEmpty();
```

Comment:
you just want to compare whether two sets are different?

set1.size() == set2.size() && set1.containsAll(set2)   is enough!

Reply:
fixed